### PR TITLE
[agent, model, perf] feat: add Gemma 3 NPU migration

### PR DESCRIPTION
**Pull request by honeyface1**
### What does this PR do?

This PR adds Gemma 3 text and VLM training support on Ascend NPU through a reproducible PatchGen specification. The implementation follows the existing Qwen3, Seed-OSS, Qwen3-VL, and Qwen3.5 NPU model integrations.

The repository diff is intentionally limited to the migration skill, Gemma 3 NPU model patch, Gemma 3 multimodal transform, Gemma 3 FLOPs accounting, runtime image-token accounting, and focused tests. No experiment YAML, launch script, dataset, log, plot, report, or checkpoint is committed. The existing Gemma 3 GPU patch and generated files are unchanged.

### Scope and acceptance status

| # | Issue #1031 criterion | Status and evidence |
|---|---|---|
| 1 | Reusable migration process | PASS. `.agents/skills/veomni-gpu-to-npu/SKILL.md` contains the model migration workflow and checklist. |
| 2 | Clean-worktree Gemma 3 workflow | PASS. PatchGen is reproducible and the commands below require no committed experiment files. |
| 3 | Ascend E2E and GPU accuracy baseline | PARTIAL. Ascend NPU text and VLM E2E pass. A Gemma 3 VLM GPU baseline is not provided in the current NPU-only scope. |
| 4 | Gemma 3 12B MFU >= 24% | NOT MET on ShareGPT4V/COCO. The best measured steady-state median is 20.04%. |
| 5 | Skill, implementation, tests, and report | PARTIAL. Skill, implementation, focused tests, commands, loss evidence, and MFU evidence are supplied here; no duplicate experiment artifacts are committed. |

This PR does not claim full completion of #1031 while criteria 3 and 4 remain unmet.

### Data provenance

The multimodal run uses [`Lin-Chen/ShareGPT4V`](https://huggingface.co/datasets/Lin-Chen/ShareGPT4V) `sharegpt4v_instruct_gpt4-vision_cap100k.json` with the corresponding COCO 2017 training images. Its `images`, `conversations`, `human/gpt`, and `<image>` record structure is handled through VeOmni's existing `sharegpt4v_sft` conversation preprocessor. Before launch, each value in the JSON `images` list must resolve to its extracted COCO image; the measured run used a local derived JSON with those paths resolved.

### Implementation

- Select the generated Gemma 3 NPU classes when Ascend is available while retaining the repository's existing GPU text path unchanged.
- Register Gemma 3 VLM as NPU-only so unsupported GPU VLM execution fails explicitly instead of reaching a missing trainer hook.
- Preserve Gemma's `1 + weight` RMSNorm semantics and dispatch RMSNorm/RoPE through NPU OpSlots with eager fallbacks.
- Construct the generated NPU text tower inside the VLM wrapper instead of resolving back to the upstream class through `AutoModel`.
- Preserve the VLM checkpoint namespaces, multimodal position IDs, `pixel_values`, `token_type_ids`, packing rules, padding-mask loss behavior, and `image_hidden_states` output.
- Add ShareGPT4V/COCO preprocessing with assistant-only labels and image-placeholder validation.
- Add Gemma 3 full/sliding-attention, SigLIP, and multimodal-projector FLOPs accounting.
- Derive Gemma 3 SigLIP token counts from `pixel_values` when grid metadata is absent so runtime MFU includes the trained vision tower.

### Verification

| Check | Result |
|---|---|
| Consecutive NPU PatchGen regeneration | PASS; generated source and diff remain byte-for-byte stable |
| Generated Python SHA-256 | `f408837ed859673b06c42b9594384991ffe997fa3a8d8bb505e3cda917267996` |
| Generated diff SHA-256 | `daa77fd27e8bbf3d6e1c0018d1de1342020dcf2ba96db43f97066b187cdabfe9` |
| Changed Python compile/import | PASS |
| Targeted Ruff check and format check | PASS |
| `pytest -q tests/utils/test_count_flops.py` | PASS, 30 tests |
| Gemma image-token runtime tests | PASS, 2 tests |
| Gemma 3 12B NPU ShareGPT4V/COCO full epoch | PASS, 3,126/3,126 steps, no runtime error |

Generate and verify the NPU model:

```bash
patchgen veomni.models.transformers.gemma3.gemma3_npu_patch_gen_config \
  -o veomni/models/transformers/gemma3/generated --diff -v
git diff --exit-code -- \
  veomni/models/transformers/gemma3/generated/patched_modeling_gemma3_npu.py \
  veomni/models/transformers/gemma3/generated/patched_modeling_gemma3_npu.diff
```

### Gemma 3 12B NPU loss

The full ShareGPT4V/COCO run used 8 Ascend 910C devices, BF16, FSDP2, DP=8, SP=1, global batch 16, micro batch 2, and a trainable vision tower.

| Metric | Result |
|---|---:|
| Optimizer steps | 3,126 / 3,126 |
| First loss | 4.28 |
| Final loss | 1.23 |
| Mean loss, first 100 steps | 2.1099 |
| Mean loss, final 100 steps | 1.1747 |
| Peak allocated memory | 49.46 GB/NPU |
| Runtime errors | none |

The loss decreases normally across the full epoch. No GPU parity claim is made for this VLM run.

### Gemma 3 12B COCO MFU

The FLOPs calculation includes the dense Gemma text tower, the 40 sliding-attention and 8 full-attention layers, the trained SigLIP vision tower, and the multimodal projector. Each 896x896 image produces `64 x 64 = 4,096` SigLIP patch tokens before projection.

| Item | Baseline measurement | Gradient-accumulation measurement |
|---|---:|---:|
| Hardware | 8 x Ascend 910C (`910_93`) | same |
| Peak BF16 throughput | 354 TFLOPS/device | same |
| Precision / parallelism | BF16, FSDP2, DP=8, SP=1 | same |
| Global / micro batch | 16 / 2 | 64 / 2 |
| Measured steps | 30 | 15 |
| Steady-state window | 11-30 | 6-15 |
| Median effective text length | 470.875 | 470.618 |
| Median step time | 1.512 s | 5.658 s |
| Median achieved throughput | 530.559 TFLOP/s aggregate | 567.402 TFLOP/s aggregate |
| Median MFU | 18.73% | **20.04%** |
| Peak allocated memory | 48.22 GB/NPU | 54.49 GB/NPU |

The larger global batch amortizes optimizer and communication overhead but does not reach the issue's 24% target. Increasing the micro batch to 3 caused severe memory pressure and reduced performance, so it is not reported as a valid result.

```text
achieved_TFLOP/s = training_FLOPs_per_step / step_seconds / 1e12
MFU = sum_rank(achieved_TFLOP/s) / (8 * 354)
```

Run the NPU COCO training/MFU measurement without committing a config or script:

```bash
source /usr/local/Ascend/ascend-toolkit/set_env.sh
ASCEND_RT_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
bash train.sh tasks/train_vlm.py configs/text/gemma3.yaml \
  --model.model_path /path/to/gemma-3-12b-it \
  --data.train_path /path/to/sharegpt4v_instruct_gpt4-vision_cap100k.json \
  --data.source_name sharegpt4v_sft \
  --data.datasets_type mapping \
  --data.chat_template tokenizer \
  --data.max_seq_len 8192 \
  --data.dataloader.type native \
  --data.dataloader.num_workers 8 \
  --model.ops_implementation.attn_implementation sdpa \
  --model.ops_implementation.rms_norm_implementation npu \
  --model.ops_implementation.rotary_pos_emb_implementation npu \
  --model.ops_implementation.swiglu_mlp_implementation eager \
  --model.ops_implementation.cross_entropy_loss_implementation npu \
  --model.accelerator.ulysses_size 1 \
  --model.accelerator.gradient_checkpointing.enable false \
  --train.global_batch_size 64 \
  --train.micro_batch_size 2 \
  --train.max_steps 30 \
  --train.wandb.enable true \
  --train.checkpoint.save_steps 0 \
  --train.checkpoint.save_epochs 0 \
  --train.checkpoint.hf_save_steps 0 \
  --train.checkpoint.hf_save_epochs 0 \
  --train.checkpoint.save_hf_weights false
```

`EnvironMeter.step()` publishes `mfu`, `flops_achieved(T)`, effective length, throughput, and memory in `step_env_metrics`; the existing WandB callback records them when enabled.

### Checklist

- [x] NPU PatchGen output is reproducible.
- [x] Gemma 3 NPU text and multimodal paths compile and import.
- [x] Gemma 3 12B ShareGPT4V/COCO completes a full NPU epoch with decreasing finite loss.
- [x] COCO MFU includes the trained vision tower and is reproducible through runtime metrics.
- [x] Existing Gemma 3 GPU patch/generated files are unchanged.
- [x] No configs, scripts, datasets, logs, plots, reports, or checkpoints are committed.
- [ ] GPU VLM accuracy baseline.
- [ ] Gemma 3 12B COCO MFU >= 24%.

### CODE REVIEW

File: .agents/skills/veomni-gpu-to-npu/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+Ascend NPU. The output is a fully functional NPU patchgen path: NPU
+patch config, generated modeling file, training config, E2E script,
+tests, and a practice report.
+
+**References (read first, load on demand):**
+
+- `docs/design/patchgen.md` — patchgen DSL, CLI, CI drift check
+- `docs/design/kernel_selection.md` — unified kernel registry, OpSlot dispatch
+- `docs/design/unified_kernel_registry.md` — per-op backend availability matrix
+- `.agents/knowledge/constraints.md` — hard constraints (never edit `generated/`, transformers v5.9.0, FSDP2)
```
[the remark this answers is not in the document]
Comment (honeyface1) [reply .agents/skills/veomni-gpu-to-npu/SKILL.md]: Resolved in the current skill: it no longer hard-codes Transformers 5.9.0 and requires the installed version to match the repository pin (currently 5.16.1), so the workflow stays correct when the repository pin advances.

File: .agents/skills/veomni-gpu-to-npu/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+
+| Model family | Weight init | Formula | NPU handling |
+|---|---|---|---|
+| Qwen3, Llama, most | `ones` | `x * rsqrt(var+eps) * weight` | `standard_rms_norm_forward_npu(x, weight, eps)` |
+| Gemma 3 | **zeros** | `x * rsqrt(var+eps) * (1.0 + weight)` | Pass `1.0 + self.weight` to OpSlot |
+| Qwen3.5 | `zeros` | `x * rsqrt(var+eps) * (1.0 + weight)` | `qwen3_5_rms_norm_forward_npu(x, 1.0+weight, eps)` |
+
+Check the model's RMSNorm class:
+```bash
+rg "class.*RMSNorm" -A 15 generated/patched_modeling_<model>_gpu.py
```
[the remark this answers is not in the document]
Comment (honeyface1) [reply .agents/skills/veomni-gpu-to-npu/SKILL.md]: Resolved in the current skill. Generated-model commands use repository-root paths under veomni/models/transformers/<model>/generated/.

File: veomni/models/transformers/gemma3/__init__.py [hunk 3]
```diff
@@ -12,16 +12,34 @@
...
 from ...loader import MODELING_REGISTRY
 
 
 @MODELING_REGISTRY.register("gemma3_text")
 def register_gemma3_text_modeling(architecture: str | None):
-    from .generated.patched_modeling_gemma3_gpu import Gemma3ForCausalLM, Gemma3TextModel
+    if IS_NPU_AVAILABLE:
+        from .generated.patched_modeling_gemma3_npu import Gemma3ForCausalLM, Gemma3TextModel
+    else:
+        from .generated.patched_modeling_gemma3_gpu import Gemma3ForCausalLM, Gemma3TextModel
```
[the remark this answers is not in the document]
Comment (honeyface1) [reply veomni/models/transformers/gemma3/__init__.py:24]: Keeping the established device-aware lazy class import used by Qwen3, Seed-OSS, Qwen3-VL, and Qwen3.5. The MODELING_REGISTRY decorators still execute when this package is imported; only generated class loading is deferred until lookup. Moving these imports to module scope would eagerly load Gemma generated modeling whenever veomni.models.transformers imports every model package.

File: .agents/skills/veomni-gpu-to-npu/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+- [ ] NPU backends verified in the live registry
+- [ ] Model-specific normalization/loss semantics preserved
+- [ ] Patch spec, generated files, and import-time dispatch implemented
+- [ ] Text and multimodal model types covered as applicable
+- [ ] Patchgen rerun is reproducible
+- [ ] Import, quality, and focused correctness checks pass
+- [ ] GPU/NPU E2E uses matched inputs and hyperparameters
+- [ ] Loss curve and numerical comparison reported
+- [ ] MFU setup, formula, peak hardware value, and result reported
+- [ ] PR diff contains no configs, scripts, logs, datasets, plots, or checkpoints unless requested
```
[the remark this answers is not in the document]
Comment (honeyface1) [reply .agents/skills/veomni-gpu-to-npu/SKILL.md:306]: The fence is already closed at line 290, before the Completion Checklist. The current file has 22 fence delimiters (an even count), so no source change is needed.

File: veomni/models/transformers/gemma3/gemma3_npu_patch_gen_config.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+        else:
+            logits = self.lm_head(hidden_states).float()
+            loss, _, fused_linear_aux = self.loss_function(
+                logits=logits,
+                labels=labels,
+                vocab_size=self.config.text_config.vocab_size,
+                hidden_states=hidden_states,
+                weights=self.lm_head.weight,
+                **lm_kwargs,
+            )
```
[the remark this answers is not in the document]
Comment (honeyface1) [reply veomni/models/transformers/gemma3/gemma3_npu_patch_gen_config.py:242]: Fixed in 77eea8c. Both eager and fused paths now receive attention-mask-filtered labels. A tiny Gemma 3 VLM regression check produced identical eager and chunk_loss values: 3.494375229.

File: .gitignore [hunk 6]
```diff
@@ -171,3 +171,9 @@ nsysoutput/
 
 # Agents worksspace should not be tracked by the primary repo.
 .agents_workspace/
+
+# NPU migration: training checkpoints, data, and debug artifacts
```
[the remark this answers is not in the document]
Comment (honeyface1) [reply .gitignore]: Obsolete after scope cleanup: .gitignore and the one-off E2E launcher are no longer part of this PR. Reproduction commands and evidence remain in the PR body.

File: docs/npu_migration/ascend_community_experience_report.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+处于 BF16 跨硬件训练的预期数值噪声范围内。
+
+### 4.5 性能（12B VLM，MFU 基准）
+
+| 指标 | 值 |
+|---|---|
+| 硬件 | Ascend 910C × 8 (NPUs 8–15) |
+| 模型 | `gemma-3-12b-it` (12B, VLM) |
+| Seq length | 8192 (dyn_bsz 打包) |
+| Micro batch | 1 / Global batch | 8 |
```
[the remark this answers is not in the document]
Comment (honeyface1) [reply docs/npu_migration/ascend_community_experience_report.md]: Obsolete after scope cleanup: the standalone report was removed from this PR; the corrected results are presented directly in the PR body.

File: run_gemma3.sh [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+export HCCL_NPU_SOCKET_PORT_RANGE=17000-18000
+
+mkdir -p logs
+
+bash train.sh tasks/train_text.py configs/text/gemma3_12b_npu.yaml \
+    --model.model_path ../models/gemma-3-12b-it \
+    --data.train_path ../data/tulu-first2000.parquet \
+    --train.accelerator.fsdp_config.fsdp_mode fsdp2 \
+    --train.init_device meta \
+    2>&1 | tee logs/gemma3_npu.log
```
[the remark this answers is not in the document]
Comment (honeyface1) [reply run_gemma3.sh]: Obsolete after scope cleanup: all one-off launch scripts were removed from this PR. The PR body provides direct shell commands instead.
